okhttp-tls/src/main/kotlin/okhttp3/tls/HeldCertificate.kt

```diff
@@ -315,88 +321,134 @@ class HeldCertificate(
+        initSign(issuerKeyPair.private)
+        update(CertificateAdapters.tbsCertificate.toDer(tbsCertificate).toByteArray())
+        sign().toByteString()
       }
-      val signatureAlgorithm = if (signedByKeyPair.private is RSAPrivateKey) {
-        "SHA256WithRSA"
-      } else {
-        "SHA256withECDSA"
+
+      // Complete signed certificate.
+      val certificate = Certificate(
+          tbsCertificate = tbsCertificate,
+          signatureAlgorithm = signatureAlgorithm,
+          signatureValue = BitString(
+              byteString = signature,
+              unusedBitsCount = 0
+          )
+      )
+
+      return HeldCertificate(subjectKeyPair, certificate.toX509Certificate())
+    }
+
+    private fun subject(): List<List<AttributeTypeAndValue>> {
+      val result = mutableListOf<List<AttributeTypeAndValue>>()
+
+      if (ou != null) {
+        result += listOf(AttributeTypeAndValue(
+            type = ObjectIdentifiers.organizationalUnitName,
+            value = ou
+        ))
       }
-      val generator = X509V3CertificateGenerator()
-      generator.setSerialNumber(serialNumber)
-      generator.setIssuerDN(signedByPrincipal)
-      generator.setNotBefore(Date(notBefore))
-      generator.setNotAfter(Date(notAfter))
-      generator.setSubjectDN(subject)
-      generator.setPublicKey(heldKeyPair.public)
-      generator.setSignatureAlgorithm(signatureAlgorithm)
+
+      result += listOf(AttributeTypeAndValue(
+          type = ObjectIdentifiers.commonName,
+          value = cn ?: UUID.randomUUID().toString()
+      ))
+
+      return result
+    }
+
+    private fun validity(): Validity {
+      val notBefore = if (notBefore != -1L) notBefore else System.currentTimeMillis()
+      val notAfter = if (notAfter != -1L) notAfter else notBefore + DEFAULT_DURATION_MILLIS
+      return Validity(
+          notBefore = notBefore,
+          notAfter = notAfter
+      )
+    }
+
+    private fun extensions(): MutableList<Extension> {
+      val result = mutableListOf<Extension>()
 
       if (maxIntermediateCas != -1) {
-        generator.addExtension(X509Extensions.BasicConstraints, true,
-            BasicConstraints(maxIntermediateCas))
+        result += Extension(
+            extnID = ObjectIdentifiers.basicConstraints,
+            critical = true,
+            extnValue = BasicConstraints(
+                ca = true,
+                pathLenConstraint = 3
+            )
+        )
       }
 
       if (altNames.isNotEmpty()) {
-        val encodableAltNames = arrayOfNulls<ASN1Encodable>(altNames.size)
-        for (i in 0 until altNames.size) {
-          val altName = altNames[i]
-          val tag = when {
-            altName.canParseAsIpAddress() -> GeneralName.iPAddress
-            else -> GeneralName.dNSName
+        val extensionValue = altNames.map {
+          when {
+            it.canParseAsIpAddress() -> {
+              generalNameIpAddress to InetAddress.getByName(it).address.toByteString()
```

Comment:
Is this known to be expanded to the full form always for IPv6?

Reply:
Yep. The result will always be 4 or 16 bytes.